### PR TITLE
AP_AdvancedFailsafe: automatically set mode to auto when datalink loss

### PR DESCRIPTION
**Pull request by eppravitra**
Automatically set mode to auto when there is datalink loss in Advanced Failsafe. This is to make sure advanced failsafe is entered when Plane is flying with other autonomous modes. For example, right now when Plane is flying in Guided mode, it will not enter the specified mission item unless the mode is switched to Auto

### CODE REVIEW

**Review by peterbarker — COMMENTED** [review]
```
8:41 PM]Ep: This would make AFS works in GUIDED as well
For example, you can have GUIDED say 5 km away. If GCS is loss while in GUIDED, AFS will not trigger
```

Taking this to DevCall, but this does look like an important fix to me.

We would want an autotest and an implementation of set_mode_auto for the other vehicles that do afs before we merge.

File: libraries/AP_AdvancedFailsafe/AP_AdvancedFailsafe.cpp
Comment (tridge) [inline libraries/AP_AdvancedFailsafe/AP_AdvancedFailsafe.cpp:243]: I think we need an AFS_OPTIONS bitmask (an AP_Int16) which has an option to enable this new behaviour. Some users may be relying on the current behaviour, eg. if they takeover with an RC switch to force QLAND mode, then find they keep going back to AUTO

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_AdvancedFailsafe/AP_AdvancedFailsafe.cpp
Comment (eppravitra) [reply libraries/AP_AdvancedFailsafe/AP_AdvancedFailsafe.cpp:243]: Thanks @tridge 
I'll get back to this PR once #23334 is sorted out. I just added AFS_OPTIONS in that PR. That PR is a bit simpler than this one.

Comment (eppravitra) [reply libraries/AP_AdvancedFailsafe/AP_AdvancedFailsafe.cpp]: AFS_OPTIONS for force entering AUTO is now added.

**Review by tridge — APPROVED** [review]

**Comment by tridge** [comment]
@eppravitra looks good to me, I'll add as dev call topic to get OK from @rmackay9

**Comment by peterbarker** [comment]
Merged, thanks!

**Comment by eppravitra** [comment]
@Hwurzburg 
Please see the proposed wiki change
https://github.com/ArduPilot/ardupilot_wiki/pull/5424

**Pull request merged**
